### PR TITLE
Added tsx generic typings for Field, Form, and FormSpy components

### PR DESCRIPTION
**Pull request by ChadLefort**
As mentioned in https://github.com/final-form/react-final-form/pull/516#discussion_r291005795, this should now allow for TSX generics for `Field`, `Form`, and `FormSpy`. This should allow you to pass in a type to either of these components, or in the case of the `Form` component if you pass in a typed `initialValues` prop typescript will infer the type.

For example:
```typescript
const initialValues: { firstName: string } = { firstName: 'chad' };

<Form initialValues={initialValues} onSubmit={onSubmit}>
  {({ values, form }) => (
    // values and form api get the proper types
  )}
</Form>
```

### CODE REVIEW

**Comment by erikras** [comment]
This pull request **introduces 1 alert** when merging 58fb5f94d9f78e8c644a67a930592ce4720309cd into 3dc413aba533fcbbc3127800e6938ea9a1ad225a - [view on LGTM.com](https://lgtm.com/projects/g/final-form/react-final-form/rev/pr-6e4da2f7963acf45266abebcb46e3324b1579226)

**new alerts:**

* 1 for Unused variable, import, function or class

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: typescript/index.d.ts [hunk 1]
```diff
@@ -64,26 +64,26 @@ export interface FormProps<FormValues = object>
...
-  format?: (value: any, name: string) => any;
+  defaultValue?: FieldValue;
+  format?: (value: FieldValue, name: string) => any;
   formatOnBlur?: boolean;
-  initialValue?: any;
+  initialValue?: FieldValue;
   isEqual?: (a: any, b: any) => boolean;
   multiple?: boolean;
-  parse?: (value: any, name: string) => any;
+  parse?: (value: FieldValue, name: string) => FieldValue;
```
Comment (erikras) [inline typescript/index.d.ts]: This `value` parameter should still be `any`, I think, as it's what is returned from the input. i.e. say we had a field of type `number`, but we had it in an `<input type="text"/>`, the whole point of the `parse` function is to do that conversion. I don't think it's a common enough use case to provide an `UnparsedValue` generic type.

Comment (ChadLefort) [reply typescript/index.d.ts]: Yea, that's true. Should have though about it a little bit more before I switched that. I'll switch it back to `any`.

**Comment by erikras** [comment]
This is AWESOME.

**Comment by ChadLefort** [comment]
Updates for `FieldArray` also:
https://github.com/final-form/react-final-form-arrays/pull/88

**Comment by erikras** [comment]
This pull request **introduces 1 alert** when merging db845a07bd9f797b40c23fd3334da5843162b90b into 3dc413aba533fcbbc3127800e6938ea9a1ad225a - [view on LGTM.com](https://lgtm.com/projects/g/final-form/react-final-form/rev/pr-a635ff0e0b53c8a491b40a92a8c21ead95821282)

**new alerts:**

* 1 for Unused variable, import, function or class

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by erikras** [comment]
Chad! My man!! 👍

**Comment by erikras** [comment]
This pull request **introduces 1 alert** when merging 730d7aa4952c2595a4890cd332243af2bad26576 into 02bf89878b1802669cbed0aacc7431a205a34fd7 - [view on LGTM.com](https://lgtm.com/projects/g/final-form/react-final-form/rev/pr-92f7f8a5bb71414ab431f4982a2cea4625e60b13)

**new alerts:**

* 1 for Unused variable, import, function or class

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by erikras** [comment]
Published in [`v6.2.0`](https://github.com/final-form/react-final-form/releases/tag/v6.2.0).

**Pull request merged**